### PR TITLE
client,daemon: add progress information to the daemon

### PR DESCRIPTION
**Pull request by mvo5**
[this includes https://github.com/ubuntu-core/snappy/pull/710]

This adds basic progress information to the `snap` command. E.g.:

```
# ./snap install cap-test.mvo
1.23 MB / 1.23 MB [========================================] 100.00 % 1.75 MB/s 
```

### CODE REVIEW

**Comment by chipaca** [comment]
Only small concern is that, because it uses a pointer to the task instead of the task itself, and then doesn't check for nil-ness before happily following the pointer.

You could make me happy by having a constructor that made sure it wasn't nil, or by embedding task directly.

Otherwise, LGTM and :+1:

File: daemon/api.go [hunk 1]
```diff
@@ -638,6 +639,35 @@ func (inst *snapInstruction) Agreed(intro, license string) bool {
...
 }
 
+func (inst *snapInstruction) Start(pkg string, total float64) {
+	inst.task.mu.Lock()
+	defer inst.task.mu.Unlock()
+
+	inst.task.total = int(total)
+}
+
+func (inst *snapInstruction) Set(current float64) {
```
Comment (zyga) [inline daemon/api.go]: Why not SetCurrent() ?

**Comment by zyga** [comment]
+1 with one nitpick on naming (the hardest problem)

**Comment by mvo5** [comment]
@chipaca Thanks for the suggestion, this is a bit tricky. The (rest)task has to be a point because we need to be able to update the value of "cur, total" for the progress reporting. So it has to be shared between the daemon (via task.Map()) and the goroutine that runs the AsyncResponse. The constructor is also tricky because the (rest)task is created only after inst is created and initialized. Creating the (rest)task earlier is possible but will require some more code changes and a split so that (rest)task creation and task running is independent.

**Comment by chipaca** [comment]
@mvo5 ah well, good enough for now i gues

**Comment by niemeyer** [comment]
retest this please

**Comment by niemeyer** [comment]
This is building a dependency on the busy loop which is knowingly a hack and already been taken down on #746. We need a different mechanism for retrieving the latest progress which is more dependable.

**Comment by niemeyer** [comment]
After looking through it in more detail, it feels like we're a bit lost on our own abstractions here. The snapInstruction has a dispatch function which busy loops waiting for the result of a change just so it can return an error code to the task which is created precisely so that we do not block the client.

I suspect we can delete the code for most of these intermediate layers once we finish the on going conversion to the state engine. Namingly, the snap instruction creates the change, and we return the change id. The client can then ask about the change and we tell it. Nothing in between.

Ideally we'd merge this and then refactor together when the time comes, but I'm concerned this is making the issue deeper. It already conflicts with simplifications coming in #746.

**Comment by mvo5** [comment]
@niemeyer Thanks for the review. I agree that the level of abstractions make this now pretty hard, I close the pull request for now and will reopen once the refactor is done, I think the approach of creating the change, returning the id and having the client ask about the change id is the right way forward.

**Pull request closed without merging**